### ISSUE TITLE
Compiler crash with tasklike when arity mismatch between types

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#16531]
csc.exe exited with code -2146232797. Arity mismatch between tasklike type and its async method builder causes an exception

Roslyn doesn't detect errors in the following code snippet but is unable to compile it. 

```cs
using System;
using System.Threading.Tasks;
using System.Runtime.CompilerServices;

public class Program
{
  public async MyAwesomeType<string> CustomTask()
  {
    await Task.Delay(1000);
    return string.Empty;
  }
}

[AsyncMethodBuilder(typeof(CustomAsyncTaskMethodBuilder<,>))]
public struct MyAwesomeType<T>
{
  public T Result { get; set; }
}

public class CustomAsyncTaskMethodBuilder<T, V>
{
  public MyAwesomeType<T> Task => default(MyAwesomeType<T>);
  public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : INotifyCompletion where TStateMachine : IAsyncStateMachine { }
  public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : ICriticalNotifyCompletion where TStateMachine : IAsyncStateMachine { }
  public static CustomAsyncTaskMethodBuilder<T, V> Create() { return default(CustomAsyncTaskMethodBuilder<T, V>); }
  public void SetException(Exception exception) { }
  public void SetResult(T t) { }
  public void SetStateMachine(IAsyncStateMachine stateMachine) { }
  public void Start<TStateMachine>(ref TStateMachine stateMachine) where TStateMachine : IAsyncStateMachine { }
}


namespace System.Runtime.CompilerServices
{
    public class AsyncMethodBuilderAttribute : System.Attribute
    {
        public AsyncMethodBuilderAttribute(Type type) { }
    }
}
```

Here is the stacktrace obtained via TryRoslyn:
```
Wrong number of type arguments 
Parameter name: typeArguments
 at Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol.Construct(ImmutableArray`1 arguments, Boolean unbound) in ...\Src\Compilers\CSharp\Portable\Symbols\NamedTypeSymbol.cs:line 876
 at Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol.ConstructWithoutModifiers(ImmutableArray`1 arguments, Boolean unbound) in ...\Src\Compilers\CSharp\Portable\Symbols\NamedTypeSymbol.cs:line 854
 at Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol.Construct(TypeSymbol[] typeArguments) in ...\Src\Compilers\CSharp\Portable\Symbols\NamedTypeSymbol.cs:line 787
 at Microsoft.CodeAnalysis.CSharp.AsyncMethodBuilderMemberCollection.TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method, TypeMap typeMap, AsyncMethodBuilderMemberCollection& collection) in ...\Src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncMethodBuilderMemberCollection.cs:line 218
 at Microsoft.CodeAnalysis.CSharp.AsyncRewriter..ctor(BoundStatement body, MethodSymbol method, Int32 methodOrdinal, AsyncStateMachine stateMachineType, VariableSlotAllocator slotAllocatorOpt, TypeCompilationState compilationState, DiagnosticBag diagnostics) in ...\Src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncRewriter.cs:line 28
 at Microsoft.CodeAnalysis.CSharp.AsyncRewriter.Rewrite(BoundStatement body, MethodSymbol method, Int32 methodOrdinal, VariableSlotAllocator slotAllocatorOpt, TypeCompilationState compilationState, DiagnosticBag diagnostics, AsyncStateMachine& stateMachineType) in ...\Src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncRewriter.cs:line 58
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt) in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1250
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1000
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 433
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass20_0.<CompileNamedTypeAsTask>b__0() in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 323
 at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in ...\Src\Compilers\Core\Portable\InternalUtilities\UICultureUtilities.cs:line 136
 at System.Threading.Tasks.Task.InnerInvoke()
 at System.Threading.Tasks.Task.Execute() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers() in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 238
 at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean generateDebugInfo, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in ...\Src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 137
 at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in ...\Src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 2340
 at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken) in ...\Src\Compilers\Core\Portable\Compilation\Compilation.cs:line 2075
 at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CancellationToken cancellationToken) in ...\Src\Compilers\Core\Portable\Compilation\Compilation.cs:line 2023
 at TryRoslyn.Core.Processing.CodeProcessor.Process(String code, ProcessingOptions options) in D:\#TeamCity\buildAgent\work\596e688e8f7f38c8\!sites\dotnet-master\Core\Processing\CodeProcessor.cs:line 44
 at TryRoslyn.Web.Api.Controllers.RoslynController.Compilation(CompilationArguments arguments) in D:\#TeamCity\buildAgent\work\596e688e8f7f38c8\!sites\dotnet-master\Web.Api\Controllers\RoslynController.cs:line 18
 at lambda_method(Closure , Object , Object[] )
 at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)
 at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)
 at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken) --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Tracing.Tracers.HttpControllerTracer.<ExecuteAsyncCore>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Tracing.ITraceWriterExtensions.<TraceBeginEndAsyncCore>d__18`1.MoveNext() --- End of stack trace from previous location where exception was thrown ---
 at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
 at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()
```

**Comment by TessenR** [comment]
I believe that AsyncMethodBuilderMemberCollection.ValidateBuilderType() shouldn't allow such arity mismatches.
Here is the current arity check from that method.
```cs
bool isArityOk = isGeneric
                 ? builderType.IsUnboundGenericType && builderType.ContainingType?.IsGenericType != true
                 : !builderType.IsGenericType;
```

**Comment by jcouv** [comment]
Tracked by VSO https://devdiv.visualstudio.com/DevDiv/_workitems?id=369804

**Comment by jcouv** [comment]
Fixed by #16608
